### PR TITLE
WIP: Document noteworthy unusual Julia features.

### PR DESCRIPTION
**Pull request by Ismael-VC**
Ref: https://github.com/JuliaLang/julia/issues/11966

Document unicode identifiers.

[av skip]

### CODE REVIEW

File: doc/manual/noteworthy-differences.rst [hunk 1]
```diff
@@ -6,6 +6,44 @@
...
+        b² = β[4]
+        Na² = β[5]
+        for i in 1:length(pixels)
+            x, y = pixels[i]
+            ret[i] = Na²*twoDGauss(x, y, μx, μy, σ) + b²
+        end
+        ret
+    end
+
+But be wary...
```
Comment (Ismael-VC) [inline doc/manual/noteworthy-differences.rst]: @ScottPJones

Comment (ScottPJones) [inline doc/manual/noteworthy-differences.rst]: Maybe add something about the two example cases that you need to be wary about (referencing a global variable thinking it was an operator applied to a local variable), and updating a variable without remembering to update the cached value.

The other example, of `/` returning a different type than it's arguments, and needing to use `div`
for integer division, along with the possibility of losing digits, should also be added.

Good that you are doing this! I wish I'd had this a couple of months ago!

File: doc/manual/noteworthy-differences.rst [hunk 2]
```diff
@@ -6,6 +6,44 @@
 Noteworthy Differences from other Languages
 *******************************************
 
+Noteworthy unusual Julia features
+---------------------------------
+
+* Julia allows identifiers with complex unicode, for example you could represent the value of a chi-squared test with ``χ²``, ``χ⁽³⁾`` is used in electromagnetism for the third-order nonlinear susceptibility and ``∇²`` is used in electromagnetism for the third-order nonlinear susceptibility. All of these unicode symbols are valid Julia identifiers.
```
Comment (jiahao) [inline doc/manual/noteworthy-differences.rst]: You didn't change the description of `∇²`, which is often used to denote the Laplacian.

File: doc/manual/noteworthy-differences.rst [hunk 3]
```diff
@@ -6,6 +6,44 @@
...
+
+    # Lennard-Jones 12-6 potential energy
+    function lennard_jones(x, A, B)
+        x⁻² = x^-2
+        x⁻⁶ = (x⁻²)^3
+        x⁻¹² = (x⁻⁶)^2
+        A*x⁻¹² - B*x⁻⁶
+    end
+
+    # Others
```
Comment (jiahao) [inline doc/manual/noteworthy-differences.rst]: cc @rsrock - can we have a better description for this code snippet?

File: doc/manual/noteworthy-differences.rst [hunk 4]
```diff
@@ -6,6 +6,44 @@
...
 
+Noteworthy unusual Julia features
+---------------------------------
+
+* Julia allows identifiers with complex unicode, for example you could represent the value of a chi-squared test with ``χ²``, ``χ⁽³⁾`` is used in electromagnetism for the third-order nonlinear susceptibility and ``∇²`` is used in electromagnetism for the third-order nonlinear susceptibility. All of these unicode symbols are valid Julia identifiers.
+
+.. doctest::
+
+    # x² instead of xx, x_squared, etc.
+    for x = 1:r; x² = x*x
```
Comment (jiahao) [inline doc/manual/noteworthy-differences.rst]: Using semicolons is discouraged

File: doc/manual/noteworthy-differences.rst (hunk 2 above)
Comment (jiahao) [inline doc/manual/noteworthy-differences.rst]: Unicode is a proper noun and should be capitalized.

Comment (Ismael-VC) [inline doc/manual/noteworthy-differences.rst]: Fixed

File: doc/manual/noteworthy-differences.rst (hunk 3 above)
Comment (Ismael-VC) [inline doc/manual/noteworthy-differences.rst]: I've just left the  L-J potential example, until more documented examples are provided or found.

File: doc/manual/noteworthy-differences.rst (hunk 4 above)
Comment (Ismael-VC) [inline doc/manual/noteworthy-differences.rst]: Fixed

File: doc/manual/noteworthy-differences.rst (hunk 1 above)
Comment (Ismael-VC) [inline doc/manual/noteworthy-differences.rst]: Done.

File: doc/manual/noteworthy-differences.rst [hunk 5]
```diff
@@ -6,6 +6,39 @@
...
+  function lj(x, A, B)
+      x⁻² = x^-2
+      x⁻⁶ = (x⁻²)^3
+      x⁻¹² = (x⁻⁶)^2
+      A*x⁻¹² - B*x⁻⁶
+  end
+
+  .. note::
+  
+  Be wary about referencing a global variable while thinking it was an operator applied to a local variable and updating a
```
Comment (IainNZ) [inline doc/manual/noteworthy-differences.rst]: I have no idea what this sentence is trying to say. Could you word it differently? Is this related to unicode?

File: doc/manual/noteworthy-differences.rst [hunk 6]
```diff
@@ -6,6 +6,39 @@
...
+  variable without remembering to update the cached value.
+
+* The ``/`` operator gives floating-point results for integer arguments, use ``div`` for truncating to an integer.
```
Comment (IainNZ) [inline doc/manual/noteworthy-differences.rst]: Should probably just be in the C differences section, people coming from Python, MATLAB, R, etc. wouldn't find this a noteworthy or unusual Julia feature.

File: doc/manual/noteworthy-differences.rst [hunk 7]
```diff
@@ -6,6 +6,39 @@
 Noteworthy Differences from other Languages
 *******************************************
 
+Noteworthy and unusual Julia features
+---------------------------------
+
+* Julia allows identifiers with complex Unicode, for example you could represent the value of a chi-squared test with ``χ²``,
```
Comment (IainNZ) [inline doc/manual/noteworthy-differences.rst]: This sentence is way too long. Also, I don't think using Unicode in identifiers in general is what would cause problems, I think its more just this one particular case.

File: doc/manual/noteworthy-differences.rst [hunk 8]
```diff
@@ -6,6 +6,39 @@
...
+  similarly ``χ⁽³⁾`` is used in electromagnetism for the third-order nonlinear susceptibility, another example is ``∇²``
+  which is the name of the Laplacian operator. All of these Unicode symbols are valid Julia identifiers.
+
+  It can also be useful for caching powers of ``x`` to ensure that intermediate products are reused optimally in time-critical code.
```
Comment (IainNZ) [inline doc/manual/noteworthy-differences.rst]: I get why this example was put here, but this is a microoptimization that is going to freak people out if its the first thing they see in a section of the manual. I'm not sure an example is needed.

File: doc/manual/noteworthy-differences.rst (hunk 6 above)
Comment (ScottPJones) [inline doc/manual/noteworthy-differences.rst]: I think it is a lot more languages than just C/C++, also Java, Lua, Python, etc.
Can you give an example of that actually working in MATLAB or R, or any other language?
AFAICT, it doesn't follow the Unicode standard for identifiers http://www.unicode.org/reports/tr31
```python
>>> x³ = 5
  File "<stdin>", line 1
    x³ = 5
     ^
SyntaxError: invalid character in identifier
>
```

Comment (IainNZ) [inline doc/manual/noteworthy-differences.rst]: @ScottPJones, its talking about division, not unicode, what are you talking about?

File: doc/manual/noteworthy-differences.rst (hunk 7 above)
Comment (ScottPJones) [inline doc/manual/noteworthy-differences.rst]: It is a whole set of characters that at least as I understand the Unicode standard UAX 31, shouldn't be considered part of identifiers.  (I have no problem at all with Unicode in identifiers, I just brought up a point where it seemed inconsistent with other uses of Unicode characters in Julia)

Comment (IainNZ) [inline doc/manual/noteworthy-differences.rst]: Great. Instead of replying to me, write that in documentation form, which is not what is here right now.

File: doc/manual/noteworthy-differences.rst (hunk 8 above)
Comment (ScottPJones) [inline doc/manual/noteworthy-differences.rst]: One would hope also that the compiler would be smart enough to lift things out of loops, if you have `x*x` or `x^2` inside a loop.  (If Jeff can't make the compiler do that, or more likely, has done it already, I'd eat my hat)

File: doc/manual/noteworthy-differences.rst (hunk 6 above)
Comment (ScottPJones) [inline doc/manual/noteworthy-differences.rst]: Never mind, I thought you were talking about the part above.

File: doc/manual/noteworthy-differences.rst (hunk 5 above)
Comment (Ismael-VC) [inline doc/manual/noteworthy-differences.rst]: @ScottPJones would you mind showing a short example of this pitfall?

File: doc/manual/noteworthy-differences.rst (hunk 6 above)
Comment (Ismael-VC) [inline doc/manual/noteworthy-differences.rst]: Ok I'll change it to the C part.

**Comment by JeffBezanson** [comment]
I find it really problematic to have a section like this in the manual. What counts as "unusual" is highly subjective. Specific items like how `/` behaves are best documented in the "differences from other languages" sections, where we can have objective descriptions. If we want to call attention to certain characters we allow in identifiers, that can go in the existing section on variable names. I don't see the purpose of hand picking a few things to isolate as "unusual" (e.g. inviting debate on whether a certain feature is unusual enough to merit inclusion in this section...).

**Comment by Ismael-VC** [comment]
@JeffBezanson what I would really want, is more people describing more differences with short examples, so that they can be documented in whatever section they end up being needed. Also I know it's subjective but Julia is indeed unusual in many ways, I don't care about the title, it could be whatever we see fit. All I want is better documentation. don't tell me that you can't come up with something unusual about Julia that I could help document better!

**Comment by Ismael-VC** [comment]
What I mean is that some features are not only noteworthy different to just one specific language, but instead are different to many languages at once, which is why some comparisons don't belong in one particular language section, also the purpose of this document is to concentrate this kind of stuff, instead of having it spread all over the manual.

**Comment by JeffBezanson** [comment]
If you know language X, then documenting differences with language X is useful. I don't understand the utility of documenting things that are considered different from nothing in particular. Who is the intended audience?

File: doc/manual/noteworthy-differences.rst (hunk 5 above)
Comment (carnaval) [inline doc/manual/noteworthy-differences.rst]: This "pitfall" is a contrived strawman crafted to make a specific point.

**Comment by Ismael-VC** [comment]
You are assuming that there must be only a one to one relationship between each Julia feature with each other language. All I want is to avoid redundancy (if a feature is related to more than one language) and centralize this kind of information (ie. the style advices are scattered) , but if you guys think it's useless feel free to close this.

**Comment by ihnorton** [comment]
The example could probably go in http://docs.julialang.org/en/latest/manual/interacting-with-julia/ if you want to split out to a new PR. Thanks!
